providers/implementations/encode_decode/decode_der2key.c

```diff
@@ -529,19 +563,246 @@ static void ecx_key_adjust(void *key, struct der2key_ctx_st *ctx)
+        if ((key = ossl_ml_kem_key_new(libctx, ctx->propq,
+                                       ctx->desc->evp_type)) != NULL
+            && ossl_ml_kem_genkey(p, p + ML_KEM_RANDOM_BYTES, NULL, 0, key))
+            ret = key;
+        else
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_GENERATE_ERROR,
+                           "error generating %s private key from seed",
+                           vinfo->algorithm_name);
+        break;
+    case ASN1_FMT:
+        if (*p++ != V_ASN1_OCTET_STRING || *p++ != 0x82)
+            break;
+        p = OSSL_CRYPTO_load_u16_be(&keylen, p);
+        if (keylen != (plen -= ML_KEM_OCTET_STRING_OVERHEAD))
+            break;
+        /* fallthrough */
+    case LONG_FMT:
+        /* Check public key consistency if provided? */
+        if ((key = ossl_ml_kem_key_new(libctx, ctx->propq,
+                                       ctx->desc->evp_type)) != NULL
+            && ossl_ml_kem_parse_private_key(p, vinfo->prvkey_bytes, key))
+            ret = key;
+        else
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY,
+                           "error parsing %s private key",
+                           vinfo->algorithm_name);
+        break;
+    }
+
+  end:
+    PKCS8_PRIV_KEY_INFO_free(p8inf);
+    if (ret == NULL)
+        ossl_ml_kem_key_free(key);
+    return ret;
+}
+
+static ossl_inline void *
+ml_kem_d2i_PUBKEY(const uint8_t **der, long der_len,
+                  struct der2key_ctx_st *ctx)
+{
+    const ML_KEM_VINFO *vinfo = ossl_ml_kem_get_vinfo(ctx->desc->evp_type);
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(ctx->provctx);
+    ML_KEM_KEY *ret = NULL;
+    BARE_PUBKEY *spki;
+    const uint8_t *end = *der;
+
+    /*
+     * Not our NID!  Is there a sensible error to report here?  Why was this
+     * decoder called?
+     */
+    if (vinfo == NULL)
+        return NULL;
+
+    /*
+     * Check that we have the right OID, the bit string has no "bits left" and
+     * that we consume all the input exactly.
+     */
+    if (der_len != ML_KEM_SPKI_OVERHEAD + (long) vinfo->pubkey_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_ENCODING,
+                       "unexpected %s public key length: %ld != %ld",
+                       vinfo->algorithm_name, der_len,
+                       ML_KEM_SPKI_OVERHEAD + (long) vinfo->pubkey_bytes);
+        return NULL;
+    }
+
+    if ((spki = OPENSSL_zalloc(sizeof(*spki))) == NULL)
+        return NULL;
+
+    /* The spki storage is freed on error */
+    if (ASN1_item_d2i_ex((ASN1_VALUE **)&spki, &end, der_len,
+                            ASN1_ITEM_rptr(BARE_PUBKEY), NULL, NULL) == NULL) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_ENCODING,
+                       "malformed %s public key ASN.1 encoding",
+                       vinfo->algorithm_name);
+        return NULL;
+    }
+
+    /* The spki structure now owns some memory */
+    if ((spki->pubkey->flags & 0x7) != 0 || end != *der + der_len) {
```

Comment:
I really dont like this magic number being used here..
As an aside it also seems like the logic is kind of wrong in ossl_asn1_string_set_bits_left().
flags gets set to 8 normally (The flag looks like it was supposed to be set if there were bits left -  but it sets it when there is none). ossl_asn1_string_get_bits_left() maybe ?

Reply:
This magic number is an essential property of ASN.1 bit strings, they have optional neglected bits, which MUST be zero in this case. If there are issues with the OpenSSL ASN.1 bitstring API, they should be addressed separately.